opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp

```diff
@@ -346,8 +358,76 @@ class EclMaterialLawManager
 
             // read the end point scaling info for the saturation region
             unscaledEpsInfo_[satnumIdx].extractUnscaled(deck, eclState, satnumIdx);
+        }
+
+        EclEpsGridProperties epsGridProperties, epsImbGridProperties;
+        epsGridProperties.initFromDeck(deck, eclState, /*imbibition=*/false);
+        if (enableHysteresis())
+            epsImbGridProperties.initFromDeck(deck, eclState, /*imbibition=*/true);
+
+        initParamsForElements_(materialLawParams_, deck, eclState, compressedToCartesianElemIdx, satnumRegionArray,
+                               gasOilUnscaledPointsVector, oilWaterUnscaledPointsVector, gasOilEffectiveParamVector,oilWaterEffectiveParamVector,
+                               epsGridProperties, epsImbGridProperties );
+
 
+        // for well connections
+        std::vector<int> satnumRegionArrayConnections;
+        std::vector<int> compressedToCartesianElemIdxConnections;
+
+        const auto& schedule = eclState.getSchedule();
+        const auto& eclGrid = eclState.getInputGrid();
+
+        size_t timeStep = 0;
+        auto wells = schedule.getWells(timeStep); //need all wells ...
```

Comment:
> Is there an alternative?

I am not sufficiently familar with the current case to answer that - but my general answer is that I think people often should try *a bit harder* before inventing a new datastructure. As I see it an ideal world should have two types of datastructures:

1. Long lived public instances which represent well defined domain quantities - *preferably immutable*. The `Completion` and `Well` objects from the parser are of this type.
2. Short lived value objects, only visible and used within a very limited scope. These can be more tailored to numerical performance/algorithm than the domain.

In particular in the well domain I feel the opm code base has structures which are intermediate between these two; where `struct Wells` and `WellState` are the main offenders - these classes/structures have way too high visibility.